docs/src/proposals/timely-vote-credits.md

```diff
@@ -0,0 +1,136 @@
+
+Historical voting data has been analyzed over many epochs and the data
+suggests that the smallest grace period that allows for very minimal impact on
+well behaved distant validators is 3 slots, which means that all slots voted
+on within 3 slots will award maximum vote credits to the voting validator.
+This gives validators nearly 2 seconds to land their votes without penalty.
+The minimum latency between two points on Earth is about 100 ms, so allowing a
+full 1,500 ms to 2,000 ms latency without penalty should not have adverse
+impact on distant validators.
+
+### Maximum Vote Credits
+
+Another factor to consider is what the maximum vote credits to award for a
+vote should be.  Assuming linear reduction in vote credits awarded (where 1
+slot of additional lag reduces earned vote credits by 1), the maximum vote
+credits value determines how much "penalty" there is for each additional slot
+of latency.  For example, a value of 10 would mean that after the grace period
+slots, every additional slot of latency would result in a 10% reduction in
+vote credits earned as each subsequent slot earns 1 credit less out of a
+maximum possible 10 credits.
+
+Again, historical voting data was analyzed over many epochs and the conclusion
+drawn was that a maximum credits of 10 is the largest value that can be used
+and still have a noticeable effect on known laggers.  Values higher than that
+result in such a small penalty for each slot of lagging that intentional
+lagging is still too profitable.  Lower values are even more punishing to
+intentional lagging; but an attempt has been made to conservatively choose the
+highest value that produces noticeable results.
+
+The selection of these values is partially documented here:
+
+https://www.shinobi-systems.com/timely_voting_proposal
+
+The above document is somewhat out of date with more recent analysis, which
+occurred in this github issue:
+
+https://github.com/solana-labs/solana/issues/19002
+
+### Method of Implementation
+
+When a Vote or VoteStateUpdate instruction is received by a validator, it will
+use the Clock sysvar to identify the slot in which that instruction has
+landed.  For any newly voted on slot within that Vote or VoteStateUpdate
+transaction, the validator will record the vote latency of that slot as
+(voted_in_slot - voted_on_slot).  It will include safety mechanisms in case
+there are bugs; if voted_on_slot is greater than or equal to voted_in_slot, a
+vote latency of 1 will be computed instead of the large value that would
+result from integer subtraction underflow.
+
+These vote latencies will be stored a new vector of u8 latency values appended
+to the end of the VoteState.  VoteState currently has ~200 bytes of free space
+at the end that is unused, so this new vector of u8 values should easily fit
+within this available space.  Because VoteState is an ABI frozen structure,
+utilizing the mechanisms for updating frozen ABI will be required, which will
```

Comment:
I think you already mentioned you were changing this, but just wanted to make a note here that we are introducing a new field to tract this instead of using the later bits in the existing votestates

Reply:
> I think you already mentioned you were changing this, but just wanted to make a note here that we are introducing a new field to tract this instead of using the later bits in the existing votestates

Agreed.  Are you saying you wanted me to add something to the proposal to call this out? I wasn't clear on whether that was your intent.

And this proposal has already been changed from the previous version where I had proposed re-using unused bits from the Lockout data structure, so we can just pretend that never happened :)

Reply:
ah sorry, misread this to mean we were using the ` ~200 bytes of free space at the end` to store the latencies. I see now you already updated to the latest proposal, thanks!

Reply:
> ah sorry, misread this to mean we were using the ` ~200 bytes of free space at the end` to store the latencies. I see now you already updated to the latest proposal, thanks!

But the proposal IS to use the ~200 bytes of free space at the end of vote accounts to store latencies.  Is that not clear?

Reply:
ah I thought it meant the ones at the end of the lockout u32's, yeah this is more repurposing the free space.